src/workerd/io/io-context.c++

```diff
@@ -1222,68 +1213,38 @@ public:
 
   kj::Promise<uint64_t> pumpTo(
       kj::AsyncOutputStream& output, uint64_t amount) override {
-    amount = co_await inner->pumpTo(output, amount);
-    quota -= amount > quota ? quota : amount;
-    co_return amount;
+    return inner->pumpTo(output, amount);
   }
 
 private:
   kj::Own<kj::AsyncInputStream> inner;
-  size_t quota;
-  kj::Own<kj::PromiseFulfiller<size_t>> fulfiller;
+  kj::Own<kj::PromiseFulfiller<void>> fulfiller;
 };
 
 }  // namespace
 
-jsg::Promise<kj::Maybe<IoOwn<kj::AsyncInputStream>>> IoContext::makeCachePutStream(
+jsg::Promise<IoOwn<kj::AsyncInputStream>> IoContext::makeCachePutStream(
     jsg::Lock& js, kj::Own<kj::AsyncInputStream> stream) {
-  KJ_IF_SOME(length, stream->tryGetLength()) {
-    // This Cache API subrequest would exceed the total PUT quota. There used to be a limit on
-    // individual puts, but now this is only used as a fast path for rejecting PUTs going beyond the
-    // total quota early.
-    if (length > initialPutQuota) {
-      return js.resolvedPromise(kj::Maybe<IoOwn<kj::AsyncInputStream>>(kj::none));
-    }
-  }
-
-  auto paf = kj::newPromiseAndFulfiller<size_t>();
-
-  KJ_DEFER(cachePutQuota = kj::mv(paf.promise));
+  auto paf = kj::newPromiseAndFulfiller<void>();
 
-  return awaitIo(js, cachePutQuota.then(
-      [fulfiller = kj::mv(paf.fulfiller), stream = kj::mv(stream)](size_t quota) mutable
-          -> kj::Maybe<kj::Own<kj::AsyncInputStream>> {
-    if (quota == 0) {
-      fulfiller->fulfill(kj::cp(quota));
-      // Total PUT quota already exceeded.
-      return kj::none;
-    }
+  KJ_DEFER(cachePutSerializer = kj::mv(paf.promise));
 
-    KJ_IF_SOME(length, stream->tryGetLength()) {
-      // PUT with Content-Length. We rely on kj-http to enforce that the expected length is
-      // respected, and can just return the new quota immediately, allowing the next PUT to start.
-      KJ_DEFER(fulfiller->fulfill(kj::cp(quota)));
-      if (quota >= length) {
-        quota -= length;
-        return kj::mv(stream);
-      } else {
-        // This Cache API subrequest would exceed the total PUT quota.
-        return kj::none;
-      }
+  return awaitIo(js, cachePutSerializer.then(
+      [fulfiller = kj::mv(paf.fulfiller), stream = kj::mv(stream)]() mutable
+          -> kj::Own<kj::AsyncInputStream> {
+    if (stream->tryGetLength() != kj::none) {
+      // PUT with Content-Length. We can just return immediately, allowing the next PUT to start.
+      KJ_DEFER(fulfiller->fulfill());
+      return kj::mv(stream);
     } else {
+      // TODO(later): With Cache streams no longer having a size limit enforced by the runtime,
+      // explore if we can clean up stream serialization too.
```

Comment:
Yes, I think we can absolutely get rid of stream serialization. Its only purpose was to guarantee a limit on the total number of bytes in a conservative way. Since we no longer impose a total limit on Content-Length requests, we can treat Transfer-Encoding: chunked requests the same.

Reply:
Sounds good – removing serialization looks non-trivial based on the amount of code used for it in api/cache.c++, but that also means it would be pretty helpful, seems to reduce code size for cache significantly too. Internal tests that check for serialization will also need to be updated. For this PR I just changed the code comments to refer to serialization and added the TODOs to keep track of this.